src/button.rs

```diff
@@ -0,0 +1,53 @@
+use rppal::gpio::{Error, Gpio, InputPin, Level};
+use std::cmp::Ordering::{Equal, Greater, Less};
+use std::time::Duration;
+use tokio::spawn;
+use tokio::sync::broadcast;
+use tokio::time::sleep;
+use tracing::{debug, error, info};
+
+pub struct Button {
+    pub receiver: broadcast::Receiver<()>,
+}
+
+impl Button {
+    const BUTTON_POLLING_MAX_COUNT: u64 = 2;
+    const BUTTON_POLLING_INTERVAL_MILLIS: u64 = 1000;
+    const EVENT_CHANNEL_CAPACITY: usize = 16;
+
+    pub fn new(pin_number: u8) -> Result<Self, Error> {
+        let pin: InputPin = Gpio::new()?.get(pin_number)?.into_input_pullup();
+        let (sender, receiver) = broadcast::channel(Self::EVENT_CHANNEL_CAPACITY);
+        spawn(async move {
+            info!(pin=?pin, "button poll task on pin {:?} started", pin);
+            let mut change_count = 0;
+            loop {
+                match pin.read() {
+                    Level::High => {
+                        debug!("button at pin {:?} not pressed", pin);
+                        change_count = 0;
+                    }
+                    Level::Low => {
+                        debug!("button at pin {:?} pressed", pin);
+
+                        match change_count.cmp(&Self::BUTTON_POLLING_MAX_COUNT) {
+                            Less => change_count += 1,
+                            Equal | Greater => {
+                                match sender.send(()) {
+                                    Ok(_) => {}
```

Comment:
Out of curiosity: Surely there's some syntactic sugar like the `?` operator that removes the need for boiler plate code like in this pretty common use case...?

Reply:
I think `map_err` and friends could simplify this a bit. However, one of the key concepts is to ensure all branches are covered and returned Results are used. So one way or the other, we need to deal with the error somewhere.

Reply:
or do you mean the match on the conditions? You could use a classic if/else construct, but i think this is even more clear and ensures we're covering all cases. I also wouldn't call it boiler plate, this is basically our main logic in this module.

Reply:
I'm talking about this particular line. The `Ok => Do nothing, Err => Do something` use case is surely a pretty common one and might be covered by some syntactic sugarcandy like the aforementioned `?` operator.

Reply:
@tmh-alex
* https://github.com/washed/drempelbox/pull/31/commits/554ccb504bb2270862c722d6937a386eb9f42362 uses `map_err`,  imo harder to read than the match expression
* https://github.com/washed/drempelbox/pull/31/commits/7d3eab198bce2541584a930fa73796ef1d569562 uses `if let Err...`, maybe a bit shorter than the match expression, ok to read

We could of course intentionally ignore the result entirely, but this is bad style.
Remember: we want the task itself to be infallible and the rust style of error handling _forces_ us to use the result and do something with it, so propagating the error upwards using `?` is not the best option as the error may be transient (channel is at capacity, for example; granted, very unlikely).

Reply:
@washed Don't intend to break good coding style and upset fellow rustaceans^^ I think the second example with `if let` is what I was looking for. Thanks a lot for both examples!

Reply:
Confirmed that it works